osc-server/src/test/java/org/osc/core/broker/service/tasks/conformance/openstack/sfc/UpdateServiceFunctionChainTaskTest.java

```diff
@@ -16,17 +16,170 @@
+    private ApiFactoryService apiFactoryServiceMock;
+
+    @Mock
+    private SdnRedirectionApi sdnApi;
+
+    @Captor
+    private ArgumentCaptor<List<NetworkElement>> neListCaptor;
+
+    @InjectMocks
+    private UpdateServiceFunctionChainTask task;
+
+    private SecurityGroup sg;
+    private ServiceFunctionChain sfc;
+    private List<NetworkElement> networkElementList;
+
+    @Before
+    public void testInitialize() throws Exception {
+        MockitoAnnotations.initMocks(this);
+
+        Mockito.when(this.em.getTransaction()).thenReturn(this.tx);
+
+        this.txControl.setEntityManager(this.em);
+
+        Mockito.when(this.dbMgr.getTransactionalEntityManager()).thenReturn(this.em);
+        Mockito.when(this.dbMgr.getTransactionControl()).thenReturn(this.txControl);
+
+        VirtualizationConnector mockVc = mock(VirtualizationConnector.class);
+        this.sg = new SecurityGroup(mockVc, "PROJECT_ID", "PROJECT_NAME");
+        this.sg.setId(2L);
+        this.sg.setName("A_SECURITY_GROUP");
+
+        Mockito.when(this.em.find(SecurityGroup.class, this.sg.getId())).thenReturn(this.sg);
+
+        this.sfc = new ServiceFunctionChain();
+        this.sfc.setId(3L);
+        this.sfc.setName("A_SFC_NAME");
+
+        Mockito.when(this.em.find(ServiceFunctionChain.class, this.sfc.getId())).thenReturn(this.sfc);
+
+        Mockito.when(this.apiFactoryServiceMock.createNetworkRedirectionApi(mockVc)).thenReturn(this.sdnApi);
+
+    }
+
     @Test
-    public void test() {
-        fail("Not yet implemented");
+    public void testExecuteTransaction_WithSecurityGroupSFCIdSetToNull_ExpectUpdate() throws Exception {
+        // Arrange
+        UpdateServiceFunctionChainTask updateTask = this.task.create(this.sfc, this.sg, this.networkElementList);
+        Mockito.when(this.sdnApi
+                .updateNetworkElement(argThat(new NetworkElementIdMatcher(this.sg.getNetworkElementId())), any()))
+                .thenReturn(new NetworkElementImpl("UPDATED_SFC_ID"));
+
+        assertNull(this.sg.getNetworkElementId());
+
+        // Act.
+        updateTask.execute();
+
+        // Assert
+        assertEquals(this.sg.getNetworkElementId(), "UPDATED_SFC_ID");
+    }
+
+    @Test
+    public void testExecuteTransaction_WithSecurityGroupSFCIdSetToNotEmpty_ExpectUpdate() throws Exception {
+        // Arrange
+        this.sg.setNetworkElementId("OLD_SFC_ID");
+        UpdateServiceFunctionChainTask updateTask = this.task.create(this.sfc, this.sg, this.networkElementList);
+        Mockito.when(this.sdnApi
+                .updateNetworkElement(argThat(new NetworkElementIdMatcher(this.sg.getNetworkElementId())), any()))
+                .thenReturn(new NetworkElementImpl("UPDATED_SFC_ID"));
+        // Act.
+        updateTask.execute();
+
+        // Assert
+        assertEquals(this.sg.getNetworkElementId(), "UPDATED_SFC_ID");
+    }
+
+    @Test
+    public void testExecuteTransaction_CallsSDNUpdate_ExpectUpdate() throws Exception {
```

Comment:
I thought all tests called SDNUpdate. What is the intent of this test? Maybe make the name match better.

Reply:
changed to CallsSDNUpdateWithProvidedParameters